packages/flutter/lib/src/material/tabs.dart

```diff
@@ -1291,36 +1297,41 @@ class _TabBarViewState extends State<TabBarView> {
 
     final int previousIndex = _controller.previousIndex;
     if ((_currentIndex - previousIndex).abs() == 1) {
-      _warpUnderwayCount += 1;
-      await _pageController.animateToPage(_currentIndex, duration: kTabScrollDuration, curve: Curves.ease);
-      _warpUnderwayCount -= 1;
+      setState(() {
+        _warpUnderwayCount += 1;
+      });
+      await _pageController.animateToPage(_currentIndex, duration: _controller.ongoingAnimationDuration ?? kTabScrollDuration, curve: _controller.ongoingAnimationCurve ?? Curves.ease);
+      if (!mounted)
+        return Future<void>.value();
+      setState(() {
+        _warpUnderwayCount -= 1;
+      });
       return Future<void>.value();
     }
 
     assert((_currentIndex - previousIndex).abs() > 1);
     final int initialPage = _currentIndex > previousIndex
         ? _currentIndex - 1
         : _currentIndex + 1;
-    final List<Widget> originalChildren = _childrenWithKey;
     setState(() {
       _warpUnderwayCount += 1;
 
-      _childrenWithKey = List<Widget>.from(_childrenWithKey, growable: false);
+      _childrenWithKey = List<Widget>.from(_originalChildrenWithKey, growable: false);
       final Widget temp = _childrenWithKey[initialPage];
       _childrenWithKey[initialPage] = _childrenWithKey[previousIndex];
       _childrenWithKey[previousIndex] = temp;
     });
     _pageController.jumpToPage(initialPage);
 
-    await _pageController.animateToPage(_currentIndex, duration: kTabScrollDuration, curve: Curves.ease);
+    await _pageController.animateToPage(_currentIndex, duration: _controller.ongoingAnimationDuration ?? kTabScrollDuration, curve: _controller.ongoingAnimationCurve ?? Curves.ease);
```

Comment:
Yes, if we call ```tabController.animateTo(1, duration: Duration(seconds: 10))``` and after 1 second we call ```tabController.animateTo(3)``` it will come in here again while tabcontroller has been animating

Reply:
In that case i think we are fine, I more worry about that we are relying on tabController to notify this method as soon as it starts animating. If not, we are risking of animation may not be in sync because we use the duration instead of the remaining time of the animation in tab controller